lib/merge.js

```diff
@@ -45,6 +45,10 @@ module.exports = internals.merge = function (target, source, options) {
         if (value &&
             typeof value === 'object') {
 
+            if (target[key] === value) {
+                continue;                                           // Can occur for shallow merges
+            }
```

Comment:
This probably also handles non-shallow cases where a shared object is merged. Eg:

```js
const a = { b: 1 };
const defaults = { a };
const obj = { a, c: 3 };

Hoek.applyToDefaults(defaults, obj);
```

Additionally, without these lines, this would throw if the shared object has enumerable properties that don't have a setter / is non-writable.